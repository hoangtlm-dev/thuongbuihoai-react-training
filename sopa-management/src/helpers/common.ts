// Flattens an array of arrays into a single array
export const flattenArray = <T>(pages: T[][]): T[] => {
  const result: T[] = [];
  pages.forEach((page) => {
    result.push(...page);
  });

  return result;
};
