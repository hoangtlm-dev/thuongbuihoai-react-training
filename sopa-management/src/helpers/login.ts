import {
  Account,
  LoginType
} from "@interfaces/account";
import { FieldError } from "react-hook-form";

export enum CheckType {
  login = 'login',
  email = 'email',
  password = 'password'
}

// Check login
export const checkAccount = (
  data?: Account[],
  email?: string,
  password?: string,
  checkType?: CheckType,
) => {
  switch (checkType) {
    case CheckType.login:
      return data?.some((user: Account) => user.email === email && user.password === password);
    case CheckType.email:
      return data?.some((user: Account) => user.email === email);
    case CheckType.password:
      return data?.some((user: Account) => user.password === password);
    default:
      throw new Error('Invalid check type');
  }
};

export const checkValidationStyles = (
  serverError: boolean,
  rules: FieldError | undefined,
  isDirty: boolean,
  type: LoginType,
) => {
  const checkIsDirty = isDirty ? type.info : type.default
  const style = serverError || rules ? type.error : checkIsDirty;

  return { style };
};