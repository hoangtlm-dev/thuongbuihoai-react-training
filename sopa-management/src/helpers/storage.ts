import { Product } from "../interfaces/product";

export const setStorage = (key: string, data: Product) => {
  localStorage.setItem(key, JSON.stringify(data));
};

export const getStorage = (key: string) => {
  return JSON.parse(localStorage.getItem(key) as any) || [];
};

export const clearStorage = () => {
  localStorage.clear();
};
