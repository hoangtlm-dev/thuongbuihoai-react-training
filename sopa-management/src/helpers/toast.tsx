import Toast from "../components/Toast";

export const showToast = (
  title: string,
  message: string,
  type: 'success' | 'error',
  duration: number
) => {
  return (
    <Toast
      title={title}
      message={message}
      type={type}
      duration={duration}
    />
  )
};
