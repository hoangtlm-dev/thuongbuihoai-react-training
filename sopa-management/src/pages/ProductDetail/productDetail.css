.detail {
  padding: var(--px-extra-normal) var(--py-medium);
  margin: 0 auto;
  width: var(--w-extra-large);
  min-height: 100vh;
}

.detail-product {
  display: flex;
}

.detail-image {
  width: var(--w-x-large);
  height: var(--h-extra-large);
}

.detail-info {
  margin-left: 30px;
}

.detail-name {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 30px;
}

.detail-size {
  margin: 30px 0;
}

.detail-text-size {
  margin-bottom: 12px;
}

.detail-rating {
  display: flex;
  align-items: center;
  width: 192px;
  justify-content: space-between;
  margin-top: 8px;
}

.detail-btn {
  margin: 0 0 24px 1px;
}

.detail-description {
  font-size: var(--text-extra-sm);
  line-height: 22px;
}
