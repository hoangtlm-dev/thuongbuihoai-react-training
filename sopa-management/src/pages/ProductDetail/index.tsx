import { memo, useCallback, useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import toast from 'react-hot-toast';
import { shallow } from 'zustand/shallow';

// Components
import Text,
{
  SizeType
} from '@components/common/Text';
import Price,
{
  PriceType
} from '@components/common/Price';
import Dropdown from '@components/common/Dropdown';
import Button,
{
  ButtonType
} from '@components/common/Button';
import Rating from '@common/Rating';
import RatingStar from '@common/Rating/RatingStar';
import Loading from '@components/common/Loading';

// Hooks
import {
  useFetchCartProduct,
  useFetchProductDetail
} from '@hooks/useQuery';
import {
  useMutationEditProductInCart,
  useMutationPostProductToCart
} from '@hooks/useMutate';

// Constants
import { SIZE } from '@constants/common';
import { CONFIRM_MESSAGE } from '@constants/validate';

// Interfaces
import { Product } from '@interfaces/product';

// Stores
import { useCartStore } from '@stores/cart';

// Styles
import './productDetail.css';

const ProductDetail = () => {
  // use useParams to get id
  const { id } = useParams();
  const [selectedValue, setSelectedValue] = useState('');

  // mutate
  const { mutate: postProduct } = useMutationPostProductToCart();
  const { mutate: putProduct } = useMutationEditProductInCart();

  const { data } = useFetchCartProduct();
  const { data: product, isLoading } = useFetchProductDetail(id);

  const {
    name,
    image,
    color,
    price,
    description,
    size
  } = product;

  const [carts, setCarts, addToCart] = useCartStore(
    (state) => [
      state.cart,
      state.setCart,
      state.addToCart
    ],
    shallow
  );

  useEffect(() => {
    if (data) {
      setCarts(data);
    }
  }, [data]);

  const handleAddToCart = () => {
    const newData = {
      ...product,
      quantity: 1,
      size: selectedValue || size
    };

    const currentCart = carts || [];
    const existingProductIndex = currentCart.findIndex(
      (item: Product) => item.id === product.id
    );

    if (existingProductIndex !== -1) {
      putProduct(
        {
          ...currentCart[existingProductIndex],
          quantity: currentCart[existingProductIndex].quantity + 1
        },
        {
          onSuccess: () => {
            addToCart(product, selectedValue || size);
          }
        }
      );
    } else {
      postProduct(newData, {
        onSuccess: () => {
          addToCart(product, selectedValue || size);
        }
      });
    }

    toast.success(CONFIRM_MESSAGE.ADD_SUCCESS);
  };

  const handleSelect = useCallback(
    (value?: string) => {
      setSelectedValue(value as string);
    },
    [selectedValue]
  );

  return (
    <div data-testid='detail' className='detail'>
      {isLoading ? (
        <Loading />
      ) : (
        <>
          <div data-testid='product-detail' className='detail-product'>
            <div className='detail-image'>
              <img src={image.url} alt={image.alt} />
            </div>
            <div className='detail-info'>
              <div className='detail-name'>
                <div className='detail-description'>
                  <Text text={name} type={SizeType.regular} />
                  <RatingStar
                    value='4.7(3205)'
                    type={SizeType.normal}
                    className='detail-rating'
                    classNameStar='rating-star'
                  />
                </div>
                <Price value={price} type={PriceType.tertiary} />
              </div>
              <Text text={`Color: ${color}`} type={SizeType.extraRegular} />
              <div className='detail-size'>
                <Text
                  text='Size'
                  className='detail-text-size'
                  type={SizeType.extraRegular}
                />
                <Dropdown
                  data={SIZE}
                  value={selectedValue || size}
                  onSetValue={(value?: string) => handleSelect(value)}
                />
              </div>
              <Button
                ariaLabel='Add to bag'
                children={`Add to bag $${price}`}
                type={ButtonType.secondary}
                className='detail-btn'
                onClick={handleAddToCart}
              />
              <Text className='detail-description' text={description} />
            </div>
          </div>
          <Rating />
        </>
      )}
    </div>
  );
};

export default memo(ProductDetail);
