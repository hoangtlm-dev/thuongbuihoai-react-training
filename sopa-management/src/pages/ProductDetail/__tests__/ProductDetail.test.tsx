// Libs
import '@testing-library/jest-dom';

// Components
import ProductDetail from '..';

// Helpers
import { renderWithRouterAndQuery } from '../../../helpers/testUtils';

describe('ProductDetail component', () => {
  test('should render ProductDetail component', () => {
    const { getByTestId } = renderWithRouterAndQuery(<ProductDetail />);
    expect(getByTestId('detail')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { container } = renderWithRouterAndQuery(<ProductDetail />);
    expect(container).toMatchSnapshot();
  });
});
