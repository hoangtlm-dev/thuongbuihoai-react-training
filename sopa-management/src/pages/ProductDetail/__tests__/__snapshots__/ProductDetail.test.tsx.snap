// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail component matches snapshot 1`] = `
<div>
  <div
    class="detail"
    data-testid="detail"
  >
    <div
      class="loading-container"
      data-testid="loading-page"
    >
      <div
        class="circle"
      />
      <div
        class="circle"
      />
      <div
        class="circle"
      />
    </div>
  </div>
</div>
`;
